internal/overlord/servstate/handlers.go

```diff
@@ -51,183 +54,528 @@ var (
 				if err != nil {
-					logger.Noticef("service %q log write failed: %v", req.Name, err)
+					logger.Noticef("service %q log write failed: %v", s.config.Name, err)
 				}
 			}
 		}()
 	}
 
-	releasePlan()
+	return nil
+}
 
-	okay := time.After(okayWait)
-	select {
-	case <-okay:
+// okayWaitElapsed is called when the okay-wait timer has elapsed (and the
+// service is considered running successfully).
+func (s *service) okayWaitElapsed() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	switch s.state {
+	case stateStarting:
+		s.started <- nil // still running fine after short duration, no error
+		s.transition(stateRunning)
+
+	default:
+		// Ignore if timer elapsed in any other state.
 		return nil
-	case <-active.done:
-		releasePlan, err := m.acquirePlan()
-		if err == nil {
-			if m.services[req.Name].cmd == cmd {
-				delete(m.services, req.Name)
+	}
+	return nil
+}
+
+// exited is called when the service's process exits.
+func (s *service) exited(waitErr error) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.resetTimer != nil {
+		s.resetTimer.Stop()
+	}
+
+	switch s.state {
+	case stateStarting:
+		code := exitCode(s.cmd)
+		s.started <- &code
+
+	case stateRunning:
+		logger.Noticef("Service %q stopped unexpectedly with code %d", s.config.Name, exitCode(s.cmd))
+		action, onType := getAction(s.config, waitErr == nil)
+		switch action {
+		case plan.ActionLog:
+			// Log has already been written above, no further log is necessary.
+			logger.Debugf("Service %q %s action is %q, transitioning to stopped state", s.config.Name, onType, action)
+			s.transition(stateStopped)
+
+		case plan.ActionExitPebble:
+			logger.Noticef("Service %q %s action is %q, triggering server exit", s.config.Name, onType, action)
+			close(s.manager.exitPebble)
+			s.transition(stateStopped)
+
+		case plan.ActionRestart:
+			backoffDurations, _ := s.config.ParseBackoff() // ignore error; it's already been validated
+			if s.backoffIndex >= len(backoffDurations) {
+				// No more backoffs, transition to stopped state.
+				logger.Noticef("Service %q %s action is %q: no more backoffs", s.config.Name, onType, action)
+				s.transition(stateStopped)
+				return nil
 			}
-			releasePlan()
+			duration := backoffDurations[s.backoffIndex]
+			logger.Noticef("Service %q %s action is %q, waiting %s before restart (backoff %d/%d)",
+				s.config.Name, onType, action, duration, s.backoffIndex+1, len(backoffDurations))
+			s.backoffIndex++
+			s.transition(stateBackoffWait)
+			time.AfterFunc(duration, func() { logError(s.backoffWaitElapsed()) })
+
+		default:
+			return fmt.Errorf("internal error: unexpected action %q", action)
 		}
-		return fmt.Errorf("cannot start service: exited quickly with code %d", cmd.ProcessState.ExitCode())
+
+	case stateTerminating, stateKilling:
+		s.stopped <- nil
+		s.transition(stateStopped)
+
+	default:
+		return fmt.Errorf("exited invalid in state %q", s.state)
 	}
-	// unreachable
+	return nil
```

Comment:
Nice logic in this function. Makes the whole thing feel well organized, thank you for the change as a whole.